File: hydra-python/src/main/python/hydra/dsl/phantom_literals.py
"""A DSL for constructing literal terms using Python's built-in datatypes."""

from hydra.phantoms import TTerm
import hydra.dsl.terms as terms

type Bigfloat = float


def string(value: str) -> TTerm[str]:
    """Construct a string term."""
    return TTerm[str](terms.string(value))


def boolean(value: bool) -> TTerm[bool]:
    """Construct a boolean term."""
    return TTerm[bool](terms.boolean(value))

def true() -> TTerm[bool]:
    """Construct a true term."""
    return boolean(True)


def false() -> TTerm[bool]:
    """Construct a false term."""
    return boolean(False)


def bigfloat(value: Bigfloat) -> TTerm[Bigfloat]:
    """Construct a bigfloat term."""
    return TTerm[Bigfloat](terms.bigfloat(value))


def bigint(value: int) -> TTerm[int]:
    """Construct a bigint term."""
    return TTerm[int](terms.bigint(value))


def binary(value: bytes) -> TTerm[bytes]:
    """Construct a binary term."""
    return TTerm[bytes](terms.binary(value))



def double(value: float) -> TTerm[float]:
    """Construct a double term."""
    return TTerm[float](terms.float64(value))


def float_(value: float) -> TTerm[float]:
    """Construct a float term."""
    return float32(value)


def float32(value: float) -> TTerm[float]:
    """Construct a float32 term."""
    return TTerm[float](terms.float32(value))


def float64(value: float) -> TTerm[float]:
    """Construct a float64 term."""
    return TTerm[float](terms.float64(value))


def integer(value: int) -> TTerm[int]:
    """Construct an int term."""
    return int32(value)


def int8(value: int) -> TTerm[int]:
    """Construct an int8 term."""
    return TTerm[int](terms.int8(value))


def int16(value: int) -> TTerm[int]:
    """Construct an int16 term."""
    return TTerm[int](terms.int16(value))


def int32(value: int) -> TTerm[int]:
    """Construct an int32 term."""
    return TTerm[int](terms.int32(value))


def int64(value: int) -> TTerm[int]:
    """Construct an int64 term."""
    return TTerm[int](terms.int64(value))

File: hydra-haskell/src/main/haskell/Hydra/Dsl/PhantomLiterals.hs
-- | A DSL for constructing literal terms using Haskell's built-in datatypes

module Hydra.Dsl.PhantomLiterals where

import Hydra.Phantoms
import qualified Hydra.Dsl.Terms as Terms

import Data.Int


-- Note: does not yet properly capture arbitrary-precision floating-point numbers,
--       because code generation does not.
type Bigfloat = Double

-- Note: does not distinguish Binary from String, because code generation does not.
type Binary = String

bigfloat :: Bigfloat -> TTerm Bigfloat
bigfloat = TTerm . Terms.bigfloat

bigint :: Integer -> TTerm Integer
bigint = TTerm . Terms.bigint

binary :: Binary -> TTerm Binary
binary = TTerm . Terms.binary

bool :: Bool -> TTerm Bool
bool = TTerm . Terms.boolean

boolean :: Bool -> TTerm Bool
boolean = bool

char :: Char -> TTerm Int
char = TTerm . Terms.char

double :: Double -> TTerm Double
double = float64

false :: TTerm Bool
false = bool False

float :: Float -> TTerm Float
float = float32

float32 :: Float -> TTerm Float
float32 = TTerm . Terms.float32

float64 :: Double -> TTerm Double
float64 = TTerm . Terms.float64

int :: Int -> TTerm Int
int = int32

int8 :: Int8 -> TTerm Int8
int8 = TTerm . Terms.int8

int16 :: Int16 -> TTerm Int16
int16 = TTerm . Terms.int16

int32 :: Int -> TTerm Int
int32 = TTerm . Terms.int32

int64 :: Int64 -> TTerm Int64
int64 = TTerm . Terms.int64

string :: String -> TTerm String
string = TTerm . Terms.string

true :: TTerm Bool
true = bool True

-- Note: untyped integers are not yet properly supported by the DSL,
--       because they are not properly supported by code generation.
uint8 :: Int8 -> TTerm Int8
uint8 = int8
uint16 :: Int16 -> TTerm Int16
uint16 = int16
uint32 :: Int -> TTerm Int
uint32 = int
uint64 :: Int64 -> TTerm Int64
uint64 = int64
